### ISSUE TITLE
How to clear deleted=true documents on couchdb

### ISSUE THREAD

**Issue by xktder** [vrtmrz/obsidian-livesync#243]
Thank you to the author for providing such an excellent plugin. Thank you.

For a question, when will the plugin on couchdb synchronize deleted documents before they are cleared? For example, this document has been marked as' deleted 'as' true'. After observing for a period of time, the document still exists and has not been cleared. When was it deleted in 'children'? This document is image data.

I have been conducting my own investigation for a period of time, but I haven't found a way to clear it. Couchdb comes with its own compression interface_ Compact seems to have no effect on this document.

I am concerned that a large number of deleted image documents have not been cleared, which will lead to excessive disk usage in couchdb.

The plugin version is 0.19.12.
The couchdb version is V3.3.2.

Thank you for the author's hard response. Thank you.

![image](https://github.com/vrtmrz/obsidian-livesync/assets/106290226/40b67147-c16f-4297-bd67-81b54e10badb)

**Comment by vrtmrz** [comment]
Thank you for asking me! 

The simplest way to remove things which are no longer needed is performing `Rebuild everything`. this operation rebuilds both the local and the remote database only with the necessary things.
(And in the initial scan, things will be stored as a bit more efficient chunk that is not split so small -- probably we need this diagram --) 

They will be removed if we enable `Do not keep metadata of deleted files` on the `Hatch` pane. If we enable that, days to delete is also can be configured.

After configured days have passed, they will actually be deleted. Then CouchDB will be able to perform compaction on the database.

However, chunks are still left there, and they waiting to be used again. Aggressive reusing chunks and differential transfers make this synchronisation speed possible. If you want to remove also them, we have to clean the database by `Clean the remote database`.

**Comment by ilmonyd** [comment]
Hi I have a question about the remote DB cleaning. I can't figure out what exactly this feature does. It says that the unused chunks are to be removed, yet it refetches them after the cleaning. Is it a bug or am i missing something. Currently I have two devices connected to the CouchDB sync: desktop and iOS, both have livesync preset selected. 

I want to remove the chunks that were created by livesync, as they expand the DB tremendously.

![image](https://github.com/vrtmrz/obsidian-livesync/assets/53193079/f2848a25-5bc5-4f5d-a480-71a1f8ee0a50)

**Comment by vrtmrz** [comment]
Thank you for asking! Yes, in the short answer, unreferenced chunks will be purged.

I am very grateful for the opportunity to explain this in more detail. (and, that is why this feature is experimental).

Self-hosted LiveSync uses two databases, the local database which stores the storage changes, and the remote database to collect and synchronise each local database.  Accordingly, there are two sorts of unreferenced chunks. Local chunks which had been no longer referenced from local documents, and remote chunks which had been no longer referenced from remote documents.

Therefore, even if our local chunks have been corresponded to existing notes after the purging local chunks, It does not mean that we will have enough chunks after the replication happened. Unfortunately, in addition to that, purged chunks will never be synchronised without being recreated. This is why the plug-in re-fetches the chunks which over-removed.
This process could not be combined. The remote chunks are possibly being used by the other clients. Due to this limitation, while in this process, the plug-in locks the remote database.And to be safe, every other device should perform some action; Fetch the database again, or also perform the GC and purge the unreferenced chunks to let they could create the chunks again.

The History of changes and chunks which had been recently referenced is important to robust synchronisation and address conflict. So we could not perform this without every device surely synchronised. This is the challenge we have to conquer in order to do GC automatically.

However,  I am also wondering about these 32 chunks are actually made for this reason. And I think also `expand the DB tremendously.` is true and truly what we have to address.
If the implementation became stable, the design and this note would be documented. However, at least I should have left the comment in the source. I am very grateful for your patience in trying out the features!

In summary, the behaviour is designed one, but possibly we have to address to improve that.

I hope this answers your question.
And let me say thank you again for reading so long note!

**Comment by ilmonyd** [comment]
Thank you for you response @vrtmrz and thank you soo much for such an amazing plugin. You are awesome for building such an amazing open source tool! 😀

**Comment by vrtmrz** [comment]
I am very relieved to hear that! And I am very honoured.
I have also written and tested some more fast GC implementations, and it looks to work better. (Still at the cost of losing history).
I will test it a bit more, and prepare to release it!
